### PR TITLE
[feat] Add MuPDF EPUB/FB2 dynamic font size

### LINKED ISSUE #4368: FR: Set font size in MuPDF EPUB rendering
* KOReader version: v2018.11.1
* Device: all

#### Issue
You can't easily adjust the font size in a MuPDF-rendered EPUB. See https://www.mobileread.com/forums/showthread.php?p=3780029#post3780029

(Tracking here so I'm less likely to forget when I have some time, and to alert potentially interested parties like @poire-z @robert00s @pazos maybe @NiLuJe ?)

It's actually very easy to tap into the backend. This works:

```diff
diff --git a/frontend/document/pdfdocument.lua b/frontend/document/pdfdocument.lua
index 34f072e9..14ae115d 100644
--- a/frontend/document/pdfdocument.lua
+++ b/frontend/document/pdfdocument.lua
@@ -42,6 +42,8 @@ function PdfDocument:init()
     if not ok then
         error(self._document)  -- will contain error message
     end
+    local Screen = Device.screen
+    self._document:layoutDocument(Screen:getWidth(), Screen:getHeight(), 50)
     self.is_open = true
     self.info.has_pages = true
     self.info.configurable = true
```

What isn't as easy, or at least what would take some time to properly implement, is to make that not a hack, to make it properly configurable and and finally to tie it into the existing font size GUI.

### PR DESCRIPTION
**Pull request by Frenzie**
Closes #4368.

<del>I might leave it like this for now (after some minor cleanup), meaning the full document will have to be reloaded, because `layoutDocument()` has to be called again. Ideally I'd give it that bit of polish too.</del> <ins>Rerendered on the fly.</ins>

### CODE REVIEW

File: frontend/apps/reader/modules/readerpaging.lua [hunk 1]
```diff
@@ -954,7 +954,8 @@ function ReaderPaging:_gotoPage(number, orig_mode)
     end
     if number > self.number_of_pages or number < 1 then
         logger.warn("wrong page number: "..number.."!")
-        return false
+        number = self.number_of_pages
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerpaging.lua]: Font size changes cause invalid page numbers. This function only protects against invalid links or something like that, not against invalid `last_page` settings. I think that's probably a more general issue, because what is you cut off a few pages from a PDF or something?

File: frontend/document/pdfdocument.lua [hunk 2]
```diff
@@ -51,6 +50,24 @@ function PdfDocument:init()
...
+    if size then
+        return size * default_font_size
+    elseif G_reader_settings:readSetting("kopt_font_size") then
+        return G_reader_settings:readSetting("kopt_font_size") * default_font_size
+    elseif DKOPTREADER_CONFIG_FONT_SIZE then
+        return DKOPTREADER_CONFIG_FONT_SIZE * default_font_size
+    else
+        return default_font_size
+    end
+    return default_font_size
```
Comment (Frenzie) [inline frontend/document/pdfdocument.lua]:
```suggestion
```

File: frontend/document/pdfdocument.lua [hunk 3]
```diff
@@ -36,12 +33,11 @@ function PdfDocument:init()
     if not ok then
         error(self._document)  -- will contain error message
     end
+    self.epub_font_size = self:getKoptFontSize()
```
Comment (poire-z) [inline frontend/document/pdfdocument.lua]: Looks ok as a whole.
Except for the namings which make it a bit hard to follow :)
`self:getKoptFontSize()`  sounds generic, but it returns something that matters only for reflowable documents.
`epub_font_size` sounds obvious, but might be better to name it `reflowable_font_size` and stay with "reflowable" in the related functions and variables names, like the nicely named `self.is_reflowable` and `document:isDocumentReflowable()`

Comment (Frenzie) [reply frontend/document/pdfdocument.lua]: Should be better now.

File: frontend/document/pdfdocument.lua [hunk 4]
```diff
@@ -36,12 +33,11 @@ function PdfDocument:init()
...
     end
+    self.reflowable_font_size = self:convertKoptToReflowableFontSize()
     -- no-op on PDF
-    self._document:layoutDocument(
-        CanvasContext:getWidth(),
-        CanvasContext:getHeight(),
-        CanvasContext:scaleBySize(self.epub_font_size))
+    self:layoutDocument()
     self.is_open = true
+    self.is_reflowable = self._document:isDocumentReflowable()
```
Comment (poire-z) [inline frontend/document/pdfdocument.lua]: nitpick: might move this one up so all the "reflowable" properties are close together

Comment (Frenzie) [reply frontend/document/pdfdocument.lua]: True, it's available right after first opening the document.

Btw, not implemented here but `fz_make_bookmark` and `fz_lookup_bookmark` should be used before & after redoing the layout to return to the same location in the reflowable document.

**Comment by poire-z** [comment]
In case you feel like fixing/enhancing on that reflowable mupdf front, it currently crashes when following footnotes link (just open any wikipedia epub), because of `link_url` being `nil` at that point:
https://github.com/koreader/koreader/blob/258b13d3443b24d7b3c8e5ba081eea42afafe6a6/frontend/apps/reader/modules/readerlink.lua#L510

Before reflowable, that code was only expecting a `page` attribute (internal links pointed to page numbers with PDF), or some external url.
(No time to dig into that, too busy learning arabic and hebrew at the moment :)

**Comment by Frenzie** [comment]
Another issue is that our code doesn't expect the number of pages to change for PDF/DjVu/CBZ (i.e., *paged* media), so you'll still have to close and reload the document after making changes — but at least you'll see an instant font size preview now.

Making the number of pages adjustable on the fly might tie into #4382, though I'm not sure if reflowing into bite-sized pages can really be done sensibly without pulling the whole document through the wringer all at once. I've got a ton of ideas with regard reflow in principle, but very little motivation. K2pdfopt can quite easily do more cool stuff, perhaps most notably working on vectors instead of bitmaps for non-scanned documents.

Basically:

a. I seldom use reflow anyway. Fit to page, page width, and content width covers close to 99 % of my use cases. That'd be up to ten percentage points lower if landscape didn't exist, I suppose.
b. What remains is very satisfying to use reflow on in case of emergency or plain straightforward documents, which is most of them, but it's much quicker to do some touch-up of any sort on a proper computer. Particularly since results are better if you run pdfsandwich (wraps around [unpaper](https://github.com/Flameeyes/unpaper) and Tesseract) or Scan Tailor  (cf. [my script](https://github.com/Frenzie/readablepdf)) first. And as soon as you're doing *that*, then it'd be silly not to run K2pdfopt afterward if desired.

The links thing should be easy to fix up though.

> (No time to dig into that, too busy learning arabic and hebrew at the moment :)

I own a book called *Cultural Connectives*. It didn't exactly teach me the Arabic script (that is, I forgot most of the basic things it said), but it's [beautifully designed](http://www.rana.im/cultural-connectives/closer_look.html).

**Comment by Frenzie** [comment]
@poire-z
> In case you feel like fixing/enhancing on that reflowable mupdf front, it currently crashes when following footnotes link (just open any wikipedia epub), because of link_url being nil at that point:

Actually I can't reproduce. If I click footnote links they direct me to a page. (And yes, I am definitely in MuPDF. :-P) Could you please give me more detailed instructions and/or a problematic (Wikipedia) EPUB?

**Comment by poire-z** [comment]
Indeed, most of the links (including the inbook TOC) work, I actually can't reproduce it at the moment on the emulator, and if I enabled "Tap to follow links" on my Kobo.

As I usually don't have "Tap to follow links enabled", it happens when holding. I have a small patch to readerhighlight to actually follow link when holding:
```diff
     elseif self.selected_word then
+        -- if hold on a link, follow link rather than dict lookup
+        if self.selected_link then
+            self:clear()
+            self.ui.link:onGotoLink(self.selected_link)
+            self.selected_word = nil
+            self.selected_link = nil
+            return true
+        end
         self:lookup(self.selected_word, self.selected_link)
```
which was probably be the cause.

But you should be able to reproduce it by holding on a link, having the dict lookup, and hiting the "Follow link" button. This I can reproduce on the emulator :):

```
08/31/19-10:46:55 DEBUG External link:
./luajit: frontend/apps/reader/modules/readerlink.lua:510: attempt to index local 'link_url' (a nil value)
stack traceback:
 frontend/apps/reader/modules/readerlink.lua:510: in function 'onGotoLink'
 frontend/ui/widget/dictquicklookup.lua:461: in function 'callback'
 frontend/ui/widget/button.lua:207: in function 'handleEvent'
 frontend/ui/widget/container/inputcontainer.lua:259: in function 'handleEvent'
 frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
 frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
 frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
 frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
```

I get some other crash on my Kobo with https://fr.wikipedia.org/wiki/P%C3%A9tra saved as EPUB, tapping in the 2.2.1 section:
```
08/31/19-10:36:29 INFO  setting zoom mode to pagewidth
./luajit: frontend/document/koptinterface.lua:739: attempt to compare number with nil
stack traceback:
    frontend/document/koptinterface.lua:739: in function 'inside_box'
    frontend/document/koptinterface.lua:746: in function 'box_distance'
    frontend/document/koptinterface.lua:759: in function 'getWordBoxIndices'
    frontend/document/koptinterface.lua:795: in function 'getTextFromBoxes'
    frontend/document/koptinterface.lua:1055: in function 'getPageBoxesFromPositions'
    frontend/apps/reader/modules/readerview.lua:481: in function 'drawPageSavedHighlight'
    frontend/apps/reader/modules/readerview.lua:467: in function 'drawSavedHighlight'
    frontend/apps/reader/modules/readerview.lua:194: in function 'paintTo'
    frontend/ui/widget/container/inputcontainer.lua:85: in function 'paintTo'
```

**Comment by Frenzie** [comment]
EPUB is a red herring here. It's the internal link logic that's broken in that scenario, just the same for PDF. It arrives as link.link.page instead of just link.page. It must've worked at some point… last touched in <https://github.com/koreader/koreader/pull/5282>.

08/31/19-11:34:43 DEBUG link {
    ["lbox"] = {
        ["y"] = 180.54998779297,
        ["x"] = 141.16000366211,
        ["h"] = 31.580017089844,
        ["w"] = 120.4700012207
    },
    ["link"] = {
        ["y1"] = 201.13000488281,l
        ["x1"] = 250.63000488281,
        ["y0"] = 191.54998779297,
        ["x0"] = 152.16000366211,
        ["page"] = 19
    },
    ["pos"] = {
        ["page"] = 5,
        ["x"] = 194.42165242165,
        ["y"] = 194.84615384615,
        ["rotation"] = 0,
        ["zoom"] = 2.3557046979866
    }
}

**Comment by Frenzie** [comment]
We can see that other code has some link.link stuff:

https://github.com/koreader/koreader/blob/258b13d3443b24d7b3c8e5ba081eea42afafe6a6/frontend/apps/reader/modules/readerlink.lua#L355-L374

https://github.com/koreader/koreader/blob/258b13d3443b24d7b3c8e5ba081eea42afafe6a6/frontend/apps/reader/modules/readerlink.lua#L376-L389

**Comment by Frenzie** [comment]
> I get some other crash on my Kobo with https://fr.wikipedia.org/wiki/P%C3%A9tra saved as EPUB, tapping in the 2.2.1 section:

I'm failing to reproduce. I tested tapping 2.2.1 in the *Sommaire* and both bootnotes in 2.2.1 itself.

**Comment by poire-z** [comment]
Well, I haven't managed to reproduce it on the emulator with a similarly looking view, but I get that crash on my kobo when I type in the red zone (not on the 2.2.1 title itself, in the text belonging to that section, sorry):
<kbd>image</kbd>

**Comment by Frenzie** [comment]
Weird. Did you create any highlights? Some pretty funny things happen there if you change the font size, though I haven't been able to trigger any crashes unless making the number of pages smaller than it was when adding the highlight.

```
08/31/19-18:10:55 DEBUG painting widget: ReaderUI
Warning in boxaSelectBySize: boxas is empty
Warning in boxaSelectBySize: boxas is empty
Error in boxaSort2d: boxas is empty
Error in pixaSort2dByIndex: naindex not defined
Error in pixaaFlattenToPixa: paa not defined
Error in pixaGetBoxa: pixa not defined
./luajit: frontend/document/koptinterface.lua:792: attempt to get length of local 'boxes' (a nil value)
stack traceback:
        frontend/document/koptinterface.lua:792: in function 'getTextFromBoxes'
        frontend/document/koptinterface.lua:1055: in function 'getPageBoxesFromPositions'
        frontend/apps/reader/modules/readerview.lua:481: in function 'drawPageSavedHighlight'
        frontend/apps/reader/modules/readerview.lua:467: in function 'drawSavedHighlight'
        frontend/apps/reader/modules/readerview.lua:194: in function 'paintTo'
        frontend/ui/widget/container/inputcontainer.lua:85: in function 'paintTo'
        frontend/ui/uimanager.lua:910: in function '_repaint'
        frontend/ui/uimanager.lua:1026: in function 'handleInput'
        frontend/ui/uimanager.lua:1098: in function 'run'
        ./reader.lua:243: in main chunk
        [C]: at 0x55a793510170
```

**Comment by poire-z** [comment]
Nope. I've been asked when opening with MuPDF to delete my bookmarks, and so they were - so Bookmarks shows no bookmark.
But looking at the metadata.epub.lua, I see some trace of highlights:
```lua
    ["highlight"] = {
        [17] = {},
        [37] = {
            [1] = {
                ["datetime"] = "2019-07-17 23:15:22",
                ["drawer"] = "lighten",
                ["pos0"] = "/body/DocFragment/body/div/ul[4]/floatBox[2]/li/div/div[2]/p/text().61",
                ["pos1"] = "/body/DocFragment/body/div/ul[5]/floatBox[1]/li/div/div[2]/p/text().24",
                ["text"] = "taille.\
L'esplanade du Deir.\
Le Deir et son esplanade"
            }
        }
    },
```
Still can't reproduce the crash with the epub and .lua taken from my Kobo on the emulator...
In case it may help, here they are: [koPetra.FR.zip](https://github.com/koreader/koreader/files/3562655/koPetra.FR.zip)

**Comment by Frenzie** [comment]
I might take a look at not crashing on invalid highlights, but only superficially. This is more of a demo feature after all. ;-) Although with these new changes it does look deceptively usable.

**Comment by poire-z** [comment]
Might be just a matter of adding `doc_settings:delSetting("highlight")` in:
https://github.com/koreader/koreader/blob/e191bf405d9f9c7390223841e4189fce8a9987b1/frontend/apps/reader/readerui.lua#L460-L471
